File: ios/Classes/EzvizManager.swift
//
//  EzvizManager.swift
//  flutter_ezviz
//
//  Created by 江鴻 on 2019/8/24.
//
import Foundation
import EZOpenSDKFramework


class EzvizManager {
    /// 获取SDK版本号
    static func sdkVersion(result: @escaping FlutterResult) {
        result(EZOpenSDK.getVersion())
    }
    
    /// 初始化SDK
    static func initSDK(_ arguments: Any?,result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let appkey = map["appkey"] as? String ?? AppKey
            let accessToken = map["accessToken"] as? String ?? AccessToken
            let enableLog = map["enableLog"] as? Bool ?? false
            let enableP2P = map["enableP2P"] as? Bool ?? false
            let ret = EZOpenSDK.initLib(withAppKey: appkey)
            EZOpenSDK.setAccessToken(accessToken)
            EZOpenSDK.setDebugLogEnable(enableLog)
            EZOpenSDK.enableP2P(enableP2P)
            EZHCNetDeviceSDK.initSDK()
            result(ret)
        }else {
            result(false)
        }
    }
    
    /// 是否开启日志
    static func enableLog(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let debug = map["enableLog"] as? Bool {
                EZOpenSDK.setDebugLogEnable(debug)
            }
        }
    }
    
    /// 是否开启P2P
    static func enableP2P(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let enableP2P = map["enableP2P"] as? Bool {
                EZOpenSDK.enableP2P(enableP2P)
            }
        }
    }
    
    /// 设置Token
    static func setAccessToken(_ arguments: Any?) {
        if let map = arguments as? Dictionary<String, Any> {
            if let accessToken = map["accessToken"] as? String {
                EZOpenSDK.setAccessToken(accessToken)
            }
        }
    }
    
    /// 获取设备信息
    static func getDeviceInfo(_ arguments: Any?, result: @escaping FlutterResult){
        if let map = arguments as? Dictionary<String, Any> {
            if let deviceSerial = map["deviceSerial"] as? String {
                EZOpenSDK.getDeviceInfo(deviceSerial) { (device, error) in
                    if let dev = device {
                        result(dev.toJSON())
                    }else {
                        ezvizLog(msg: error?.localizedDescription ?? "")
                        result(nil)
                    }
                }
            }
        }else {
            result(nil)
        }
    }
    
    /// 获取设备信息列表
    static func getDeviceInfoList(result: @escaping FlutterResult) {
        EZOpenSDK.getDeviceList(0, pageSize: 100) { (devices, count, error) in
            if let devList = devices {
                var devArr : [[String: Any]] = []
                devList.forEach({ (item) in
                    if let dev = item as? EZDeviceInfo {
                        devArr.append(dev.toJSON())
                    }
                })
                result(devList)
            }else {
                ezvizLog(msg: error?.localizedDescription ?? "")
                result(nil)
            }
        }
    }
    
    /// 设置视频通道分辨率
    static func setVideoLevel(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let deviceSerial = map["deviceSerial"] as? String ?? ""
            let cameraId = map["cameraId"] as? Int ?? 0
            let videoLevel = map["videoLevel"] as? Int ?? 0
            let ezvizVideoLevel = EZVideoLevelType.init(rawValue: videoLevel) ?? EZVideoLevelType.low
            
            EZOpenSDK.setVideoLevel(deviceSerial, cameraNo: cameraId, videoLevel: ezvizVideoLevel) { (error) in
                if let err = error {
                    ezvizLog(msg: err.localizedDescription)
                    result(false)
                }else {
                    result(true)
                }
            }
        
        }
    }
    
    /// 云台控制
    static func controlPTZ(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let deviceSerial = map["deviceSerial"] as? String ?? ""
            let cameraId = map["cameraId"] as? Int ?? 0
            let command = map["command"] as? String ?? ""
            let action = map["action"] as? String ?? ""
            let speed = map["speed"] as? Int ?? 0
            
            if let cmd = PTZKeys[command] as? EZPTZCommand{
                if let act = PTZKeys[action] as? EZPTZAction {
                    EZOpenSDK.controlPTZ(deviceSerial, cameraNo: cameraId, command: cmd, action: act, speed: speed) { (error) in
                        if let err = error {
                            ezvizLog(msg: err.localizedDescription)
                            result(false)
                        }else {
                            result(true)
                        }
                    }
                    
                }else {
                    ezvizLog(msg: "action字符串不合法")
                    result(false)
                }
            }else {
                ezvizLog(msg: "command字符串不合法")
                result(false)
            }
        }
    }
}


// MARK: - 网络设备相关操作
extension EzvizManager {
    
    /// 登录网络设备
    static func loginNetDevice(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? String ?? ""
            let pwd = map["pwd"] as? String ?? ""
            let ipAddr = map["ipAddr"] as? String ?? ""
            let port = map["port"] as? Int ?? 0
            
            DispatchQueue.global().async {
                let deviceInfo = EZHCNetDeviceSDK.loginDevice(withUerName: userId, pwd: pwd, ipAddr: ipAddr, port: port)
                
                DispatchQueue.main.async {
                    if let dev = deviceInfo {
                        result(dev.toJSON())
                    }else {
                        ezvizLog(msg: "无法登陆设备")
                        result(nil)
                    }
                }
            }
        }
    }
    
    /// 登出网络设备
    static func logoutNetDevice(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? Int ?? 0
            result(EZHCNetDeviceSDK.logoutDevice(withUserId: userId))
        }
    }
    
    /// 网络设备云台控制
    static func netControlPTZ(_ arguments: Any?, result: @escaping FlutterResult) {
        if let map = arguments as? Dictionary<String, Any> {
            let userId = map["userId"] as? Int ?? 0
            let channelNo = map["channelNo"] as? Int ?? 0
            let command = map["command"] as? String ?? ""
            let action = map["action"] as? String ?? ""
            
            if let cmd = netPTZKeys[command] as? EZPTZCommandType{
                if let act = netPTZKeys[action] as? EZPTZActionType {
                    DispatchQueue.global().async {
                        let ret = EZHCNetDeviceSDK.ptzControl(withUserId: userId, channelNo: channelNo, command: cmd, action: act)
                        DispatchQueue.main.async {
                            result(ret)
                        }
                    }
                    
                }else {
                    ezvizLog(msg: "action字符串不合法")
                    result(false)
                }
            }else {
                ezvizLog(msg: "command字符串不合法")
                result(false)
            }
        }
    }
    
}

File: android/src/main/kotlin/net/jayson/flutter_ezviz/EzvizManager.kt
package net.jayson.flutter_ezviz

import com.videogo.exception.BaseException
import com.videogo.openapi.EZConstants
import com.videogo.openapi.EZHCNetDeviceSDK
import com.videogo.openapi.EZOpenSDK
import io.flutter.plugin.common.MethodChannel.Result
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread

/**
 * 萤石云SDK管理类
 */
object EzvizManager {

    val ptzKeys = mapOf(
            Pair(Action_START,EZConstants.EZPTZAction.EZPTZActionSTART.name),
            Pair(Action_STOP,EZConstants.EZPTZAction.EZPTZActionSTOP.name),
            Pair(Command_Left,EZConstants.EZPTZCommand.EZPTZCommandLeft.name),
            Pair(Command_Right,EZConstants.EZPTZCommand.EZPTZCommandRight.name),
            Pair(Command_Up,EZConstants.EZPTZCommand.EZPTZCommandUp.name),
            Pair(Command_Down,EZConstants.EZPTZCommand.EZPTZCommandDown.name),
            Pair(Command_ZoomIn,EZConstants.EZPTZCommand.EZPTZCommandZoomIn.name),
            Pair(Command_ZoomOut,EZConstants.EZPTZCommand.EZPTZCommandZoomOut.name)
    )

    /// 获取SDK版本号
    fun sdkVersion(result: Result) {
        result.success(EZOpenSDK.getVersion())
    }

    /// 初始化SDK
    fun initSDK(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val appKey = map["appKey"] as? String ?: AppKey
            val accessToken = map["accessToken"] as? String ?: AccessToken
            val enableLog = map["enableLog"] as? Boolean ?: false
            val enableP2P = map["enableP2P"] as? Boolean ?: false
            val application = ApplicationUtils.application
            application?.let {
                val ret = EZOpenSDK.initLib(it, appKey)
                EZOpenSDK.enableP2P(enableP2P)
                EZOpenSDK.showSDKLog(enableLog)
                EZOpenSDK.getInstance().setAccessToken(accessToken)
                EZHCNetDeviceSDK.getInstance()
                result.success(ret)
            }
        }
    }

    /// 是否开启Log
    fun enableLog(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val debug = map["enableLog"] as? Boolean ?: false
            EZOpenSDK.showSDKLog(debug)
        }
    }

    /// 是否开启P2P
    fun enableP2P(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val debug = map["enableP2P"] as? Boolean ?: false
            EZOpenSDK.enableP2P(debug)
        }
    }

    /// 设置Token
    fun setAccessToken(arguments: Any?) {
        val map = arguments as? Map<*, *>
        map?.let {
            val accessToken = map["accessToken"] as? String ?: ""
            EZOpenSDK.getInstance().setAccessToken(accessToken)
        }
    }

    /// 获取设备信息
    fun getDeviceInfo(arguments: Any?, result: Result) {
        doAsync {
            val map = arguments as? Map<*, *>
            map?.let {
                try {
                    val deviceSerial = map["deviceSerial"] as? String ?: ""
                    val deviceInfo = EZOpenSDK.getInstance().getDeviceInfo(deviceSerial)
                    if (deviceInfo == null) {
                        uiThread {
                            result.success(null)
                        }
                    } else {
                        val device = mapOf<String, Any>(
                                Pair("deviceSerial", deviceInfo.deviceSerial),
                                Pair("deviceName", deviceInfo.deviceName),
                                Pair("isSupportPTZ", deviceInfo.isSupportPTZ),
                                Pair("cameraNum", deviceInfo.cameraNum)
                        )
                        uiThread {
                            result.success(device)
                        }
                    }
                } catch (e: BaseException) {
                    result.error("获取设备异常", e.localizedMessage, null)
                }
            }
        }
    }

    /// 获取设备信息List
    fun getDeviceList(result: Result) {
        doAsync {
            try {
                val deviceInfoList = EZOpenSDK.getInstance().getDeviceList(0, 100)
                if (deviceInfoList == null || deviceInfoList.isEmpty()) {
                    uiThread {
                        result.success(null)
                    }
                } else {
                    val array = arrayListOf<Map<String, Any>>()
                    for (deviceInfo in deviceInfoList) {
                        val device = mapOf<String, Any>(
                                Pair("deviceSerial", deviceInfo.deviceSerial),
                                Pair("deviceName", deviceInfo.deviceName),
                                Pair("isSupportPTZ", deviceInfo.isSupportPTZ),
                                Pair("cameraNum", deviceInfo.cameraNum)
                        )
                        array.add(device)
                    }
                    uiThread {
                        result.success(array)
                    }
                }
            } catch (e: BaseException) {
                result.error("获取设备异常", e.localizedMessage, null)
            }

        }
    }

    /// 设置视频通道分辨率
    fun setVideoLevel(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val deviceSerial = map["deviceSerial"] as? String ?: ""
            val cameraId = map["cameraId"] as? Int ?: 0
            val videoLevel = map["videoLevel"] as? Int ?: 0

            doAsync {
                val ret = EZOpenSDK.getInstance().setVideoLevel(deviceSerial, cameraId, videoLevel)
                uiThread {
                    result.success(ret)
                }
            }
        }
    }

    /// 云台控制
    fun controlPTZ(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            try {
                val deviceSerial = map["deviceSerial"] as? String ?: ""
                val cameraId = map["cameraId"] as? Int ?: 0
                val command = map["command"] as? String ?: ""
                val action = map["action"] as? String ?: ""
                val speed = map["speed"] as? Int ?: 0
                doAsync {
                    val ret = EZOpenSDK.getInstance().controlPTZ(deviceSerial, cameraId, EZConstants.EZPTZCommand.valueOf(ptzKeys[command] ?: "") , EZConstants.EZPTZAction.valueOf(ptzKeys[action] ?: ""), speed)
                    uiThread {
                        result.success(ret)
                    }
                }
            } catch (e: BaseException) {
                result.error("云台控制异常", e.localizedMessage, null)
            }
        }
    }

    /// 登录网络设备
    fun loginNetDevice(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            try {
                val userId = map["userId"] as? String ?: ""
                val pwd = map["pwd"] as? String ?: ""
                val ipAddr = map["ipAddr"] as? String ?: ""
                var port = map["port"] as? Int ?: 0

                doAsync {
                    val netDevice = EZHCNetDeviceSDK.getInstance().loginDeviceWithUerName(userId, pwd, ipAddr, port)
                    if (netDevice == null) {
                        uiThread {
                            result.success(null)
                        }
                    } else {
                        val deviceInfo = mapOf<String, Any>(
                                Pair("userId", netDevice.loginId),
                                Pair("channelCount", netDevice.byChanNum),
                                Pair("startChannelNo", netDevice.byStartChan),
                                Pair("dStartChannelNo", netDevice.byStartDChan),
                                Pair("dChannelCount", netDevice.byIPChanNum),
                                Pair("byDVRType", netDevice.byDVRType)
                        )
                        uiThread {
                            result.success(deviceInfo)
                        }
                    }
                }
            } catch (e: BaseException) {
                result.error("登录网络设备异常", e.localizedMessage, null)
            }

        }
    }

    /// 登出网络设备
    fun logoutNetDevice(arguments: Any?, result: Result) {
        val map = arguments as? Map<*, *>
        map?.let {
            val userId = map["userId"] as? Int ?: 0
            doAsync {
                val ret = EZHCNetDeviceSDK.getInstance().logoutDeviceWithUserId(userId)
                uiThread {
                    result.success(ret)
                }
            }
        }
    }

    /// 网络设备控制
    fun netControlPTZ(arguments: Any?, result: Result) {
        result.error("Android不支持此方法",null,null)
    }

}